### PR TITLE
Top file enhancements

### PR DESCRIPTION
**Pull request by cachedout**
Refs #12483

This is designed to make it easier to manage the currently undefined and effectively broken handling of multiple top files across mutiple environments.

This PR introduces some new configuration options: `top_file_merging_strategy` and `env_order`. They are both documented in the updated reference page.

This is currently a work-in-progress and will sit in the PR queue for a while so that people can provide feedback before we get it merged.

### CODE REVIEW

File: salt/state.py [hunk 1]
```diff
@@ -2308,10 +2309,17 @@ def _get_envs(self):
...
         Pull the file server environments out of the master options
         '''
-        envs = set(['base'])
+        envs = ['base']
         if 'file_roots' in self.opts:
-            envs.update(list(self.opts['file_roots']))
-        return envs.union(set(self.client.envs()))
+            envs.extend(list(self.opts['file_roots']))
+        client_envs = self.client.envs()
+        if self.opts.get('env_order', []):
```
Comment (thatch45) [inline salt/state.py]: We should put a check here to not include envs that do not exist in the return from the self.client.envs() call

**Comment by cachedout** [comment]
Note: Add `top_file_merging_strategy` to mopts!

**Comment by mathom** [comment]
Hey, I just tried out the `top_file_merge: False` option but it seems to still be merging my topfiles:
```
    Detected conflicting IDs, SLS IDs need to be globally unique.
    The conflicting ID is 'consul-service' and is found in SLS 'base:consul.install' and SLS 'pr_23:consul.install'
```

My topfiles look like:
```
{{env}}:
    '*':
        - common
    'consul-*':
        - common
        - consul.cluster
```

**Comment by arthurlogilab** [comment]
Thanks for this implementation, we bumped into  #12483 and from our point of view the ```top_file_merging_strategy: False``` would be our approach (modularity and simplicity). We'll try to find some time to test this on our setup to give you more informed feedback. 

Reading some comments on #12483 this fix seems important.

**Comment by cachedout** [comment]
Go Go Jenkins!

**Comment by cachedout** [comment]
@mathom Sorry for the big delay in responding! That config value was actually incorrect and based on a previous approach I had, which I forgot to remove. I've updated the default config with the proper value. `top_file_merging_strategy: same`, should accomplish I think you're seeking.

File: doc/ref/states/top.rst [hunk 2]
```diff
@@ -64,304 +134,224 @@ just the state tree, then the :conf_master:`file_roots` option can be expanded:
...
+In the above, we declare three environments: `dev`, `qa` and `prod`.
+Each environment has a single directory assigned to it.
+
+Our top file references the environments:
 
 .. code-block:: yaml
 
     dev:
-      'webserver*dev*':
+      'webserver*':
```
Comment (whiteinge) [inline doc/ref/states/top.rst]: These hostnames with the env contained within the ID were probably intentional. If you don't pin a minion to an env with the methods you mention below in "Choosing an Environment" then the environment is chosen from the top file. The result of this change means the `webserver*` minions are now in all the dev, qa, and prod environments!

**Comment by whiteinge** [comment]
Wrapping one's head around top file compilation is headache-inducing; it's very possible I'm doing something wrong. Question about `env_order`. Given the config:

```yaml
environment: staging

env_order:
  - staging
  - base
```

Shouldn't the top and bottom results be identical?

```bash
% salt 'kevin' state.show_top saltenv=base
kevin:
    ----------
    base:
        - wait

% salt 'kevin' state.show_top saltenv=staging
kevin:
    ----------
    staging:
        - foo

% salt 'kevin' state.show_top
kevin:
    ----------
    staging:
        - foo
```

File: doc/ref/states/top.rst (hunk 2 above)
Comment (cachedout) [inline doc/ref/states/top.rst]: Good point. I ended up removing these because I felt like they were making the documentation too difficult to understand.

**Comment by cachedout** [comment]
@whiteinge First match wins, not the last match. Should it be the other way around? First match made sense to me, but maybe I'm being silly. :-/

**Comment by whiteinge** [comment]
I tried both orders and got the same result both ways.

**Comment by cachedout** [comment]
@whiteinge Ah, K. Thanks. Investigating.

**Comment by rallytime** [comment]
Note to self: Should go back to 2015.8.

**Pull request merged**